helion/autotuner/ucb_pattern_search.py

```diff
@@ -0,0 +1,392 @@
+from .effort_profile import PATTERN_SEARCH_DEFAULTS
+from .pattern_search import PatternSearch
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator, Sequence
+
+    from ..runtime.config import Config
+    from ..runtime.kernel import BoundKernel
+
+try:
+    from botorch.acquisition import UpperConfidenceBound
+    from botorch.fit import fit_gpytorch_mll
+    from botorch.models import MixedSingleTaskGP
+    from gpytorch.mlls import ExactMarginalLogLikelihood
+
+    HAS_BO_DEPS = True
+except ImportError:
+    HAS_BO_DEPS = False
+
+
+class UCBPatternSearch(PatternSearch):
+    """
+    Upper Confidence Bound (UCB) Pattern Search
+
+    Modifies PatternSearch to (1) generate random neighbors from each search copy
+    within a set radius, (2) filter the neighbors to benchmark using a fitted GaussianProcess
+    with the UCB acquisition function.
+
+    Uses the MixedSingleTaskGP model from botorch, which supports continuous
+    and categorical variables. It only fits the GP once to avoid long runtimes.
+    """
+
+    def __init__(
+        self,
+        kernel: BoundKernel,
+        args: Sequence[object],
+        *,
+        initial_population: int = PATTERN_SEARCH_DEFAULTS.initial_population,
+        copies: int = PATTERN_SEARCH_DEFAULTS.copies,
+        max_generations: int = PATTERN_SEARCH_DEFAULTS.max_generations,
+        min_improvement_delta: float = 0.001,
+        frac_selected: float = 0.3,
+        num_neighbors: int = 100,
+        radius: int = 2,
+        ucb_beta: float = 2.0,
+    ) -> None:
+        super().__init__(
+            kernel=kernel,
+            args=args,
+            initial_population=initial_population,
+            copies=copies,
+            max_generations=max_generations,
+            min_improvement_delta=min_improvement_delta,
+        )
+        # Storage for BO
+        self.num_neighbors = num_neighbors
+        self.radius = radius
+        self.ucb_beta = ucb_beta
+
+        # Initialize config encoder
+        self.frac_selected = frac_selected
+
+        self.cat_dims = []
+        offset = 0
+        for spec in self.config_gen.flat_spec:
+            n_dims = spec.encode_dim()
+            if spec.is_categorical():
+                # All dimensions of this encoder are categorical
+                self.cat_dims.extend(range(offset, offset + n_dims))
+            offset += n_dims
+
+    def fit_gp(
+        self, train_X: torch.Tensor, train_Y: torch.Tensor, cat_dims: list
+    ) -> MixedSingleTaskGP:
+        # Filter out rows where train_Y contains inf or nan
+
+        if HAS_BO_DEPS:
+            gp = MixedSingleTaskGP(
+                train_X,
+                -train_Y.unsqueeze(-1),
+                cat_dims,
+            )
+
+            with torch.enable_grad():
+                mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
+                fit_gpytorch_mll(mll)
+
+            return gp
+        else:
+            return None
```

Comment:
Throw and `helion.exc.*` error if someone tries to use this algorithm without the deps that includes the command to run to install the deps.